src/charm.py

```diff
@@ -58,7 +58,7 @@ def __init__(self, framework: ops.Framework):
         super().__init__(framework)
 
         self.port = 2111
-
+        self.unit.set_ports(self.port)
```

Comment:
What does this do exactly? I could access gocert from https://<unit-ip>:2111 from both the VM and my main machine through an ssh tunnel.

Reply:
set_ports will patch the k8s service to open the ports we provide it, `2111` in this instance.

## Example

I have two gocert's deployed, the one with the `-k8s` suffix is from main  and the one with the `-port` port suffix is this branch here that sets ports using `self.unit.set_ports()` .

```
guillaume@thinkpad:~/code/gocert-k8s-operator$ juju status
Model  Controller          Cloud/Region        Version  SLA          Timestamp
eee    microk8s-localhost  microk8s/localhost  3.5.1    unsupported  07:11:00-04:00

App          Version  Status  Scale  Charm       Channel      Rev  Address         Exposed  Message
gocert-k8s            active      1  gocert-k8s  latest/edge    1  10.152.183.193  no       
gocert-port           active      1  gocert-k8s                 0  10.152.183.192  no       

Unit            Workload  Agent  Address      Ports  Message
gocert-k8s/0*   active    idle   10.1.19.173         
gocert-port/0*  active    idle   10.1.19.143  
```

Here only `gocert-port` is accessible at its app address `https://10.152.183.192:2111`

You can realise this also by looking at the k8s service:

```
guillaume@thinkpad:~/code/gocert-k8s-operator$ kubectl get svc -n eee
NAME                                 TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)     AGE
gocert-k8s                           ClusterIP   10.152.183.193   <none>        65535/TCP   6m41s
gocert-k8s-endpoints                 ClusterIP   None             <none>        <none>      6m41s
gocert-port                          ClusterIP   10.152.183.192   <none>        2111/TCP    4m59s
gocert-port-endpoints                ClusterIP   None             <none>        <none>      4m59s
modeloperator                        ClusterIP   10.152.183.148   <none>        17071/TCP   14h
self-signed-certificates             ClusterIP   10.152.183.23    <none>        65535/TCP   14h
self-signed-certificates-endpoints   ClusterIP   None             <none>        <none>      14h
```

Here `gocert-k8s` gets a random port (65535) and `gocert-port` gets the port we want (2111).

Reply:
makes sense